src/transformers/modeling_bart.py

```diff
@@ -935,12 +943,269 @@ def forward(
+            input_ids = input_ids.unsqueeze(1).expand(batch_size, num_return_sequences, cur_len)
+            input_ids = input_ids.contiguous().view(
+                batch_size * num_return_sequences, cur_len
+            )  # shape: (batch_size * num_return_sequences, cur_len)
+            batch_size *= num_return_sequences
+
+        # Below here somewhat similar to PretrainedModel._generate_beam_search
+        # Expand input to num beams
+        input_ids = input_ids.unsqueeze(1).expand(batch_size, num_beams, cur_len)
+
+        input_ids = input_ids.contiguous().view(batch_size * num_beams, cur_len)  # (batch_size * num_beams, cur_len)
+        if attention_mask is not None:
+            attention_mask = (
+                attention_mask.unsqueeze(1)
+                .expand(batch_size, num_beams, cur_len)
+                .contiguous()
+                .view(batch_size * num_beams, cur_len)
+            )  # RESHAPE
+        assert bos_token_id == 0
+
+        # generated hypotheses
+        finalized_hyps = [  # they end in EOS and we wont work on them more!
+            BeamHypotheses(num_beams, max_length, length_penalty, early_stopping=True) for _ in range(batch_size)
+        ]
+
+        # scores for each sentence in the beam
+        beam_scores = torch.zeros((batch_size, num_beams), dtype=torch.float, device=input_ids.device)
+        beam_scores[:, 1:] = -1e9  # avoid ties in first step
+        beam_scores = beam_scores.view(-1)  # shape (batch_size * num_beams,)
+
+        # decoder tokens
+        prev_output_tokens = input_ids.new(batch_size * num_beams, 1).long().fill_(-1)
+        prev_output_tokens[:, 0] = 2  # HARDCODED EOS, which will be removed at the end.
+        decoder_cache = None
+        done = [False for _ in range(batch_size)]  # done sentences
+
+        self.model.decoder.generation_mode = True  # tells decoder not to use causal mask
+        for step in range(max_length + 1):
+            decoder_input_ids = prev_output_tokens.clone()
+            model_inputs = self.prepare_inputs_for_generation(
+                input_ids, decoder_cache, decoder_input_ids, attention_mask,
+            )
+            outputs = self(**model_inputs)
+            lprobs = F.log_softmax(outputs[0][:, -1, :], dim=-1)
+
+            assert pad_token_id is not None, "should break circleci"
+            lprobs[lprobs != lprobs] = -math.inf  # kill nans
+            lprobs[:, pad_token_id] = -math.inf
+            # TODO(SS): unk is 3? fairseq also takes out every step, and has unk at slot 3
+
+            assert bos_token_id == 0, "configurable bos_token_id not yet supported"
+            if step == max_length:  # FORCE EOS to be chosen
+                lprobs[:, :eos] = -math.inf
+                lprobs[:, eos + 1 :] = -math.inf
+            elif step == 0:  # Force BOS to be chosen
+                lprobs[:, bos_token_id + 1 :] = -math.inf
+            if step < min_len:
+                lprobs[:, eos] = -math.inf
```

Comment:
would change the order of the if, elif, if statements here to correspond to its sequential logic: 

```
            if step == 0:  # Force BOS to be chosen
                lprobs[:, bos_token_id + 1 :] = -math.inf
            if step < min_len:
                lprobs[:, eos] = -math.inf
            elif step == max_length:  # FORCE EOS to be chosen
                lprobs[:, :eos] = -math.inf
                lprobs[:, eos + 1 :] = -math.inf          
```

Reply:
Also I'm wondering whether instead of forcing the `bos_token_id` and `eos_token_id` to be chosen, 
wouldn't it be cleaner to just start with the `bos_token_id` (also starting at step == 1 then) and finishing with the `eos_token_id` (whereas I would actually argue that the `eos_token_id` should not be appended to an unfinished sentence since the sentence did not yet reach its EOS)?